lib/monitorable_resource.rb

```diff
@@ -41,11 +41,24 @@ def process_event_loop
   def check_pulse
     @run_event_loop = true if @resource.needs_event_loop_for_pulse_check?
 
+    stale_retry = false
     @pulse_check_started_at = Time.now
     begin
       @pulse = @resource.check_pulse(session: @session, previous_pulse: @pulse)
       Clog.emit("Got new pulse.") { {got_pulse: {ubid: @resource.ubid, pulse: @pulse}} } if (rpt = @pulse[:reading_rpt]) && (rpt < 6 || rpt % 5 == 1) || @pulse[:reading] != "up"
     rescue => ex
+      if !stale_retry &&
+          (
+            # Seen when sending on a broken connection.
+            ex.is_a?(IOError) && ex.message == "closed stream" ||
+            # Seen when receiving on a broken connection.
+            ex.is_a?(Errno::ECONNRESET) && ex.message.start_with?("Connection reset by peer - recvfrom(2)")
+          ) &&
+          @session[:last_pulse]&.<(Time.now - 8)
+        stale_retry = true
+        @session.merge!(@resource.init_health_monitor_session)
```

Comment:
At any point, `@session` can be set to `nil` by the pulse thread calling `close_resource_session`. If `@session` is `nil` here (very rare, but still possible), this raises a NoMethodError, which causes the related monitor thread to exit.

I think we need this:

```suggestion
        @session&.merge!(@resource.init_health_monitor_session)
```

Probably need to add nocov because I don't think it's possible to hit this case in tests with the current code.

Note that this was in issue in the previous code, but it would be caught by this rescue. As you are inside the block, it will no longer be caught by this rescue.

An alternative approach is to wrap this code in a separate begin/rescue block.